lib/linked_list.rb

```diff
@@ -12,120 +12,251 @@ def initialize(value, next_node = nil)
-    # insert the new node at the beginning of the linked list
-    def add_first(value)
-      raise NotImplementedError
+    until current.nil?
+      return true if current.data == value
+      current = current.next
     end
+    return false
+  end
 
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    def search(value)
-      raise NotImplementedError
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  def find_max
+    if (@head == nil)
+      return nil
+    end
+    current = @head
+    temp_max = @head.data
+    until current.nil?
+      if current.data > temp_max
+        temp_max = current.data
+      end
+      current = current.next
     end
+    return temp_max
+  end
 
-    # method to return the max value in the linked list
-    # returns the data value and not the node
-    def find_max
-      raise NotImplementedError
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  def find_min
+    if @head.nil?
+      return nil
+    end
+    current = @head
+    temp_min = @head.data
+    until current.nil?
+      if current.data < temp_min
+        temp_min = current.data
+      end
+      current = current.next
     end
+    return temp_min
+  end
 
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    def find_min
-      raise NotImplementedError
+  # method that returns the length of the singly linked list
+  def length
+    count = 0
+    current = @head
+    while current != nil
+      count += 1
+      current = current.next
     end
+    return count
+  end
 
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  def get_at_index(index)
+    temp = @head
+    count = 0
+    while count < index && temp
+      temp = temp.next
+      count += 1
+    end
+    if (count == index)
+      return temp.data
+    else
+      return nil
+    end
+  end
 
-    # method that returns the length of the singly linked list
-    def length
-      raise NotImplementedError
+  # method to print all the values in the linked list
+  def visit
+    current = @head
+    until current.nil?
+      puts current.data
+      current = current.next
     end
+  end
+
+  # method to delete the first node found with specified value
+  def delete(value)
+    return nil if @head.nil?
+
+    current = @head
+    prev = nil
 
-    # method that returns the value at a given index in the linked list
-    # index count starts at 0
-    # returns nil if there are fewer nodes in the linked list than the index value
-    def get_at_index(index)
-      raise NotImplementedError
+    until current.data == value
+      prev = current
+      current = current.next

```

Comment:
Consider if the value isn't in the list.

```suggestion
      current = current.next
      return if current.nil?
```